### PR TITLE
[orchagent]: Extend the SRv6Orch to support the programming of the L3Adj

### PR DESCRIPTION
**Pull request by cscarpitta**
**What I did**
Extended the SRv6Orch to support the programming of the L3Adj associated with SRv6 uA, End.X, uDX4, uDX6, End.DX4, and End.DX6 behaviors.
HLD: https://github.com/sonic-net/SONiC/pull/1472

**Why I did it**
The current implementation of SRv6Orch does not process the L3Adj.

**How I verified it**
Added new test cases to test_srv6.py

### CODE REVIEW

**Comment by ahsalam** [comment]
@prsunny @kperumalbfn could you please help and review the code of this PR?
This feature is tracked for 202311 release and documented by this HLD https://github.com/sonic-net/SONiC/pull/1472
@zhangyanzhao

**Comment by prsunny** [comment]
> @prsunny @kperumalbfn could you please help and review the code of this PR? This feature is tracked for 202311 release and documented by this HLD [sonic-net/SONiC#1472](https://github.com/sonic-net/SONiC/pull/1472) @zhangyanzhao

@kperumalbfn to review/signoff

File: orchagent/srv6orch.cpp [hunk 1]
```diff
@@ -422,6 +422,175 @@ bool Srv6Orch::mySidExists(string my_sid_string)
...
+    {
+        /*
+         * It's a neighbor add notification, let's walk through the list of SRv6 MySID entries
+         * that are waiting for that neighbor to be ready, and install them into the ASIC.
+         */
+
+        SWSS_LOG_INFO("Neighbor ADD event: %s, installing pending SRv6 SIDs",
+                        update.entry.ip_address.to_string().c_str());
+
+        auto it = m_pendingSRv6MySIDEntries.find(NextHopKey(update.entry.ip_address.to_string()));
```
Comment (shuaishang) [inline orchagent/srv6orch.cpp:455]: The update.entry should contains ip_address + alias，you can't lookup m_pendingSRv6MySIDEntries with only ip_address.
Two neighbors may has same ip_address and different alias.

File: tests/test_srv6.py [hunk 2]
```diff
@@ -124,15 +163,173 @@ def test_mysid(self, dvs, testlog):
...
+        assert status == True
+        for fv in fvs:
+            if fv[0] == "SAI_MY_SID_ENTRY_ATTR_NEXT_HOP_ID":
+                assert fv[1] == next_hop_ipv4_id
+            if fv[0] == "SAI_MY_SID_ENTRY_ATTR_ENDPOINT_BEHAVIOR":
+                assert fv[1] == "SAI_MY_SID_ENTRY_ENDPOINT_BEHAVIOR_DX4"
+            elif fv[0] == "SAI_MY_SID_ENTRY_ATTR_ENDPOINT_BEHAVIOR_FLAVOR":
+                assert fv[1] == "SAI_MY_SID_ENTRY_ENDPOINT_BEHAVIOR_FLAVOR_PSP_AND_USD"
+
+        # create MySID END.DX6
```
Comment (kperumalbfn) [inline tests/test_srv6.py:290]: can you add a testcase that tests neighbor addition after adding mysid entry

File: orchagent/srv6orch.cpp [hunk 3]
```diff
@@ -422,6 +422,175 @@ bool Srv6Orch::mySidExists(string my_sid_string)
...
+    {
+        /*
+         * It's a neighbor add notification, let's walk through the list of SRv6 MySID entries
+         * that are waiting for that neighbor to be ready, and install them into the ASIC.
+         */
+
+        SWSS_LOG_INFO("Neighbor ADD event: %s, installing pending SRv6 SIDs",
+                        update.entry.ip_address.to_string().c_str());
+
+        auto it = m_pendingSRv6MySIDEntries.find(NextHopKey(update.entry.ip_address.to_string(), update.entry.alias.c_str()));
```
Comment (kperumalbfn) [inline orchagent/srv6orch.cpp:455]: m_pendingSRv6MySIDEntries key is only NextHopKey during insert. Could you check if the lookup will be successful. Also, please add a testcase for this scenario.

File: orchagent/srv6orch.cpp [hunk 4]
```diff
@@ -422,6 +422,175 @@ bool Srv6Orch::mySidExists(string my_sid_string)
...
+            it = tmp;
+
+            SWSS_LOG_INFO("SID %s removed successfully", my_sid_string.c_str());
+
+            /*
+             * Finally, add the SID to the pending MySID entries set, so that we can re-install it 
+             * when the neighbor comes back
+             */
+            auto pending_mysid_entry = make_tuple(my_sid_string, dt_vrf, adj, end_action);
+            m_pendingSRv6MySIDEntries[NextHopKey(update.entry.ip_address.to_string())].insert(pending_mysid_entry);
```
Comment (kperumalbfn) [inline orchagent/srv6orch.cpp:569]: NextHopKey with ip_address and interface?? Check all the other places where m_pendingSRv6MySIDEntries dictionary is accessed.

File: orchagent/srv6orch.cpp [hunk 5]
```diff
@@ -529,6 +714,36 @@ bool Srv6Orch::createUpdateMysidEntry(string my_sid_string, const string dt_vrf,
...
+        sai_object_id_t next_hop_id;
+        nexthop = NextHopKey(adj); 
+        SWSS_LOG_INFO("Adjacency %s", adj.c_str());
+        if (m_neighOrch->hasNextHop(nexthop))
+        {
+            SWSS_LOG_INFO("Nexthop for adjacency %s exists in DB", adj.c_str());
+            next_hop_id = m_neighOrch->getNextHopId(nexthop);
+            if(next_hop_id == SAI_NULL_OBJECT_ID)
+            {
+              SWSS_LOG_ERROR("Failed to get nexthop for adjacency %s", adj.c_str());
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply orchagent/srv6orch.cpp]: @eddieruan-alibaba Agree. I changed the code to add the SID to the pending list: https://github.com/sonic-net/sonic-swss/pull/2902/commits/b4c557fb8a5f68129df1bb1ca78a65649cf03806

File: orchagent/srv6orch.cpp (hunk 4 above)
Comment (cscarpitta) [reply orchagent/srv6orch.cpp]: @kperumalbfn Fixed.

File: tests/test_srv6.py (hunk 2 above)
Comment (cscarpitta) [reply tests/test_srv6.py:317]: @kperumalbfn I added the test case as requested: https://github.com/sonic-net/sonic-swss/pull/2902/commits/2e9485ce288f3a7c75fdb198e713623b4f0044f9

File: orchagent/srv6orch.cpp [hunk 6]
```diff
@@ -529,6 +714,36 @@ bool Srv6Orch::createUpdateMysidEntry(string my_sid_string, const string dt_vrf,
         attributes.push_back(vrf_attr);
         vrf_update = true;
     }
+    sai_attribute_t nh_attr;
+    NextHopKey nexthop;
+    bool nh_update = false;
+    if (mySidNextHopRequired(end_behavior))
+    {
+        sai_object_id_t next_hop_id;
+        nexthop = NextHopKey(adj);
```
[the remark this answers is not in the document]
Comment (cscarpitta) [reply orchagent/srv6orch.cpp:733]: @kperumalbfn I added a check to return an error if the adj is ECMP: https://github.com/sonic-net/sonic-swss/pull/2902/commits/6bb9fe2823de558e611492c1a26a89cb175b224d

File: orchagent/srv6orch.cpp (hunk 1 above)
Comment (cscarpitta) [reply orchagent/srv6orch.cpp]: @shuaishang I changed the code to use ip_address + alias as key when performing lookup.

**Comment by cscarpitta** [comment]
@kperumalbfn @eddieruan-alibaba @shuaishang Many thanks for the review. All comments have been addressed.

**Review by kperumalbfn — APPROVED** [review]
LGTM

**Comment by ahsalam** [comment]
Hi @prsunny, the PR has been approved by @kperumalbfn . It is ready for merge, could you please merge it ?

**Comment by cscarpitta** [comment]
Hi @prsunny This PR has been reviewed and approved. It is ready for merge. Could you please merge the PR?

**Review by prsunny — APPROVED** [review]

**Pull request merged**